controllers/openstackdataplanenodeset_controller.go

```diff
@@ -254,28 +214,34 @@ func (r *OpenStackDataPlaneRoleReconciler) Reconcile(ctx context.Context, req ct
 	// all our input checks out so report InputReady
 	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.InputReadyMessage)
 
+	// Check if nodes need to be provisioned and set the relevant BaremetalHost field.
+	nodeSetManagedHostMap := deployment.ManagedHostMap{}
+	err = deployment.BuildBMHHostMap(ctx, helper, instance, &nodeSetManagedHostMap)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	hostMap, ok := nodeSetManagedHostMap.HostMap[instance.ObjectMeta.Labels["openstackdataplane"]]
+	if ok {
+		bmsTemplate := instance.Spec.BaremetalSetTemplate.DeepCopy()
+		bmsTemplate.BaremetalHosts = hostMap
+		instance.Spec.BaremetalSetTemplate = *bmsTemplate
```

Comment:
We should not update the instance spec in the controller. Either it should be in the webhook or we update the openstackbaremetalset spec when creating/patching it.

Reply:
@rabi thanks, I agree we shouldn't be updating the spec with the controller. Kicking this can a bit further down the road to the webhook discussion independent of this PR, so I just moved it to be under the `CreateOrPatch` for the baremetal set. Does that look right to you now?

https://github.com/bshephar/dataplane-operator/blob/crd-proposal/pkg/deployment/baremetal.go#L63-L69

I'm not sure it's overly necessary for us to update `instance.Spec.BaremetalSetTemplate.BaremetalHosts` in our CR right. The main thing is that we take the existing values from `instance.Spec.BaremetalSetTemplate` and merge it with the values we set for `bmsTemplate.BaremetalHosts`. So we can probably eventually forget about this line:
https://github.com/bshephar/dataplane-operator/blob/crd-proposal/pkg/deployment/baremetal.go#L67

And just set all of the values on the `baremetalSet.Spec`. 

But I just copied the existing logic and moved it into the `CreateOrUpdate` for the BaremetalSet which I guess should satisfy the concerns here?

Reply:
We don't need to copy all existing logic there. `baremetalSet.BaremetalHosts = hostMap` would be enough. Also we we've to change `BuildBMHHostMap()` to build the map for 'a' nodeSet unlike earlier where it was doing for all roles/nodesets in the dataplane. `ManagedHostMap` has to change to `map[string]baremetalv1.InstanceSpec` as well.

Reply:
Actually, that doesn't work because it doesn't update the BaremetalHosts section prior to:
https://github.com/bshephar/dataplane-operator/blob/crd-proposal/controllers/openstackdataplanenodeset_controller.go#L218

Reply:
Yes, you need to check if the preProvisioned flag instead here. That's why I suggested  webhook as the preferred option.

Reply:
What if we instead base that check on whether `preProvisioned` is defined:
https://github.com/bshephar/dataplane-operator/blob/crd-proposal/controllers/openstackdataplanenodeset_controller.go#L218